src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/AnonymousTypeManager.Templates.cs

```diff
@@ -189,13 +189,45 @@ private NamedTypeSymbol ConstructAnonymousDelegateImplementationSymbol(Anonymous
             // If all parameter types and return type are valid type arguments, construct
             // the delegate type from a generic template. Otherwise, use a non-generic template.
             bool useUpdatedEscapeRules = Compilation.SourceModule.UseUpdatedEscapeRules;
-            if (allValidTypeArguments(useUpdatedEscapeRules, typeDescr))
+            if (allValidTypeArguments(useUpdatedEscapeRules, typeDescr, out var needsIndexedName))
             {
                 var fields = typeDescr.Fields;
                 Debug.Assert(fields.All(f => hasDefaultScope(useUpdatedEscapeRules, f)));
 
-                bool returnsVoid = fields.Last().Type.IsVoidType();
+                bool returnsVoid = fields[^1].Type.IsVoidType();
                 int nTypeArguments = fields.Length - (returnsVoid ? 1 : 0);
+                var typeArgumentsBuilder = ArrayBuilder<TypeWithAnnotations>.GetInstance(nTypeArguments);
+                for (int i = 0; i < nTypeArguments; i++)
+                {
+                    typeArgumentsBuilder.Add(fields[i].TypeWithAnnotations);
+                }
+                var typeArguments = typeArgumentsBuilder.ToImmutableAndFree();
+
+                // Delegate name cannot be fully determined by its signature (e.g., it has default parameter values).
+                if (needsIndexedName)
+                {
+                    Debug.Assert(nTypeArguments != 0);
+
+                    // Replace field types (where type parameters will be) by placeholders.
+                    var genericFieldTypes = IndexedTypeParameterSymbol.Take(nTypeArguments);
+                    if (returnsVoid)
+                    {
+                        var genericFieldTypeBuilder = ArrayBuilder<TypeWithAnnotations>.GetInstance(fields.Length);
+                        genericFieldTypeBuilder.AddRange(genericFieldTypes);
+                        genericFieldTypeBuilder.Add(fields[^1].TypeWithAnnotations);
+                        genericFieldTypes = genericFieldTypeBuilder.ToImmutableAndFree();
+                    }
+                    var genericTypeDescr = typeDescr.WithNewFieldsTypes(genericFieldTypes);
+
+                    var key = new SynthesizedDelegateKey(genericTypeDescr);
+                    var namedTemplate = this.AnonymousDelegates.GetOrAdd(key, static (key, arg) =>
+                    {
+                        var (@this, genericTypeDescr) = arg;
+                        return new AnonymousDelegateTemplateSymbol(@this, genericTypeDescr);
+                    }, (this, genericTypeDescr));
+                    return namedTemplate.Construct(typeArguments);
```

Comment:
>return namedTemplate.Construct(typeArguments); [](http://example.com/codeflow?start=20&length=46)

Consider adding an empty line before this statement. The previous statement is a multiline statement, uses a block, and its last line is not indented. At least for me, it takes an effort to "pull" statements apart while reading the code.

Reply:
Thanks, changed.